nmdc_server/auth.py

```diff
@@ -100,56 +219,139 @@ async def admin_required(user: models.User = Depends(login_required)) -> models.
+    redirect_uri in the ORCID /authorize request. The ORCID API will redirect the user back to this
+    route with an authorization code. We use this code to request an ORCID access token. The ORCID
+    access token response also includes the authenticated user's ORCID iD and name. We use this
+    information to create or update a user in our database. Finally, we generate our own
+    authorization code and redirect the user back to the original redirect_uri (which was provided
+    to the /auth/login route and stored in the session).
+    """
+    token_response = await oauth2_client.orcid.authorize_access_token(request)
+    # TODO: use token_response.access_token to make requests to the ORCID API for more user info?
+    # Or possibly store the access token in the database for later use?
+    user = User(orcid=token_response["orcid"], name=token_response["name"])
+    user_model = crud.get_or_create_user(db, user)
+    user_model.name = user.name
+    db.commit()
+    redirect_uri = request.session.pop("redirect_uri")
+    state = request.session.pop("state", None)
+    # TODO: also include code_challenge and code_challenge_method in this request to emulate PKCE?
+    code = crud.create_authorization_code(db, user_model, redirect_uri)
+    url = f"{redirect_uri}?code={code.code}"
+    if state is not None:
+        url += f"&state={state}"
+    return RedirectResponse(url)
 
-class LoginBehavior(str, Enum):
-    web = "web"
-    jwt = "jwt"
-    app = "app"
 
+@router.post("/token", response_model=TokenResponse, include_in_schema=False)
+async def token(
+    body: TokenRequest,
+    db: Session = Depends(get_db),
+):
+    """Exchange an nmdc-server authorization code for nmdc-server tokens.
 
-# authentication
-@router.get("/login", include_in_schema=False)
-async def login_via_orcid(request: Request, behavior: LoginBehavior = LoginBehavior.web):
-    qs = f"?behavior={behavior}"
-    if settings.host:
-        redirect_uri = f"{settings.host.rstrip('/')}/token{qs}"
-    else:
-        redirect_uri = request.url_for("token") + qs
-    return await oauth2_client.orcid.authorize_redirect(request, redirect_uri)
+    When a user receives an authorization code from query parameters in the redirect_uri that they
+    provided to the /auth/login route, they can exchange it for an nmdc-server access token using
+    this route. The provided authorization code is looked up in the database. If it is missing,
+    already exchanged, wasn't generated from the same redirect_uri, or is older than 5 minutes, an
+    error is raised. Otherwise, the code is marked as exchanged and new tokens are generated for the
+    user associated with the code.
+    """
+    authorization_code = crud.get_authorization_code(db, body.code)
+    if (
+        authorization_code is None
+        or authorization_code.exchanged
+        or authorization_code.redirect_uri != body.redirect_uri
+        or authorization_code.created < datetime.utcnow() - timedelta(minutes=5)
```

Comment:
Is 5 minutes the tightest reasonable delta to check for with the code exchange? Would a shorter time make a meaningful impact on security?

Reply:
I can't recall exactly what I was referencing when I chose 5 minutes. Taking a look at this again I see that https://www.oauth.com/oauth2-servers/authorization/the-authorization-response/ says:

> The OAuth 2.0 spec recommends a maximum lifetime of 10 minutes, but in practice, most services set the expiration much shorter, around 30-60 seconds.

We might as well start with something much shorter and extend it if we find we need to. Will update.